From 6c5b643a66b54650b822d25939cdbe4c5f9238c3 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@nxp.com>
Date: Wed, 7 Mar 2018 11:15:23 +0800
Subject: [PATCH 3438/5242] MLK-17156-5: ASoC: imx-pcm-rpmsg: register rpmsg
 codec

commit  6f5875b42895b03adde08ff47e195fd222d5e190 from
https://source.codeaurora.org/external/imx/linux-imx.git

register rpmsg codec after the rpmsg-audio-channel is
ready.

Signed-off-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Reviewed-by: Daniel Baluta <daniel.baluta@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/imx-pcm-rpmsg.c |   21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/sound/soc/fsl/imx-pcm-rpmsg.c b/sound/soc/fsl/imx-pcm-rpmsg.c
index 6cdc424..80a00a4 100644
--- a/sound/soc/fsl/imx-pcm-rpmsg.c
+++ b/sound/soc/fsl/imx-pcm-rpmsg.c
@@ -449,7 +449,6 @@ static int imx_rpmsg_pcm_new(struct snd_soc_pcm_runtime *rtd)
 
 int imx_rpmsg_platform_register(struct device *dev)
 {
-
 	struct fsl_rpmsg_i2s  *rpmsg_i2s = dev_get_drvdata(dev);
 
 	i2s_info_g	=  &rpmsg_i2s->i2s_info;
@@ -473,14 +472,14 @@ static int i2s_rpmsg_cb(struct rpmsg_device *rpdev, void *data, int len,
 	if (msg->header.type == I2S_TYPE_C) {
 		if (msg->header.cmd == I2S_TX_PERIOD_DONE) {
 			spin_lock_irqsave(&i2s_info_g->lock[0], flags);
-			rpmsg = &i2s_info_g->send_msg[0];
+			rpmsg = &i2s_info_g->send_msg[RPMSG_AUDIO_TX];
 			rpmsg->param.buffer_tail++;
 			rpmsg->param.buffer_tail %= i2s_info_g->num_period[0];
 			spin_unlock_irqrestore(&i2s_info_g->lock[0], flags);
 			i2s_info_g->callback[0](i2s_info_g->callback_param[0]);
 		} else if (msg->header.cmd == I2S_RX_PERIOD_DONE) {
 			spin_lock_irqsave(&i2s_info_g->lock[1], flags);
-			rpmsg = &i2s_info_g->send_msg[1];
+			rpmsg = &i2s_info_g->send_msg[RPMSG_AUDIO_RX];
 			rpmsg->param.buffer_tail++;
 			rpmsg->param.buffer_tail %= i2s_info_g->num_period[1];
 			spin_unlock_irqrestore(&i2s_info_g->lock[1], flags);
@@ -496,12 +495,28 @@ static int i2s_rpmsg_cb(struct rpmsg_device *rpdev, void *data, int len,
 
 static int i2s_rpmsg_probe(struct rpmsg_device *rpdev)
 {
+	struct platform_device *codec_pdev;
+	struct fsl_rpmsg_i2s *rpmsg_i2s = NULL;
+	int ret;
+
 	i2s_info_g->rpdev = rpdev;
 
 	init_completion(&i2s_info_g->cmd_complete);
 
 	dev_info(&rpdev->dev, "new channel: 0x%x -> 0x%x!\n",
 			rpdev->src, rpdev->dst);
+
+	rpmsg_i2s = container_of(i2s_info_g, struct fsl_rpmsg_i2s, i2s_info);
+
+	codec_pdev = platform_device_register_data(&rpmsg_i2s->pdev->dev,
+			RPMSG_CODEC_DRV_NAME, PLATFORM_DEVID_NONE,
+			NULL, 0);
+	if (IS_ERR(codec_pdev)) {
+		dev_err(&rpdev->dev, "failed to register rpmsg audio codec\n");
+		ret = PTR_ERR(codec_pdev);
+		return ret;
+	}
+
 	return 0;
 }
 
-- 
1.7.9.5

